### ISSUE TITLE
cell-var-from-loop false positive on generators

### ISSUE THREAD

**Issue by shakhat** [PyCQA/pylint#2978]
#### Rule cell-var-from-loop works correctly with list comprehension:
```
$ cat test_bad.py 
lst = [lambda: i for i in range(10)]
print([a() for a in lst])

$ python3 test_bad.py 
[9, 9, 9, 9, 9, 9, 9, 9, 9, 9]

$ pylint test_bad.py 
************* Module test_bad
test_bad.py:1:15: W0640: Cell variable i defined in loop (cell-var-from-loop)
```
#### But it gives false positive in case of generator:
```
$ cat test_good.py 
gen = (lambda: i for i in range(10))
print([a() for a in gen])

$ python3 test_good.py 
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

$ .tox/style/bin/pylint test_good.py 
************* Module test_good
test_good.py:1:15: W0640: Cell variable i defined in loop (cell-var-from-loop)
```
### Expected behavior

It's expected that no warning is produced in the latter case.

### pylint --version output
```
pylint 2.3.1
astroid 2.2.5
Python 3.6.8 (default, Jan 14 2019, 11:02:34) 
[GCC 8.0.1 20180414 (experimental) [trunk revision 259383]]
```

**Comment by SamyCookie** [comment]
Does https://github.com/PyCQA/pylint/pull/2276 fixed this issue ?

**Comment by shakhat** [comment]
No, it does not fix, #2276 is already in the version that I'm using (2.3.1).
Also I checked on the current master (298a22d9), the behaviour is the same.

**Comment by PCManticore** [comment]
Thanks, I can confirm the issue on `master`.

**Comment by PCManticore** [comment]
Looked a bit into this. While case is alright, if the generator is previously consumed by a sequence, the bug still reproduces, e.g. the following will yield `[9, 9, 9, ...]`
```
gen = (lambda: i for i in range(10))
print([a() for a in list(gen)])
[9, 9, 9, ...]
```

Given this, I'm inclined to close this issue, as it's a problematic case that might or might not result in a bug at runtime, depending on how the generator gets consumed.

**Comment by shakhat** [comment]
Agree with @PCManticore - it is safer to avoid the case, because the runtime behavior depends on how *client* code uses the generator and improper use leads to hard bugs.
